runtime/metrics.go

```diff
@@ -1,171 +1,177 @@
 }
 
-func (*Metrics) NewSummary(o SummaryOpts) Summary {
-	m := prometheus.NewSummary(o)
-	switch err := prometheus.Register(m); err.(type) {
-	case nil, prometheus.AlreadyRegisteredError:
-	default:
-		panic(err)
+func (m *Metrics) Close() error {
+	var err *multierror.Error
+
+	// close diagnostics; this stops the ticker and any further observations on
+	// runenv.D() will fail/panic.
+	err = multierror.Append(err, m.diagnostics.Close())
+
+	// close results; no more results via runenv.R() can be recorded.
+	err = multierror.Append(err, m.results.Close())
+
+	if m.influxdb != nil {
+		// Next, we reopen the results.out file, and write all points to InfluxDB.
+		results := filepath.Join(m.re.TestOutputsPath, "results.out")
+		if file, errf := os.OpenFile(results, os.O_RDONLY, 0666); errf == nil {
+			err = multierror.Append(err, m.batchInsertInfluxDB(file))
+		} else {
+			err = multierror.Append(err, errf)
+		}
 	}
-	return m
-}
 
-func (*Metrics) NewCounterVec(o CounterOpts, labels ...string) *CounterVec {
-	m := prometheus.NewCounterVec(o, labels)
-	switch err := prometheus.Register(m); err.(type) {
-	case nil, prometheus.AlreadyRegisteredError:
-	default:
-		panic(err)
+	// Flush the immediate InfluxDB writer.
+	if m.batcher != nil {
+		err = multierror.Append(err, m.batcher.Close())
 	}
-	return m
+
+	// Now we're ready to close InfluxDB.
+	if m.influxdb != nil {
+		err = multierror.Append(err, m.influxdb.Close())
+	}
+
+	return err.ErrorOrNil()
 }
 
-func (*Metrics) NewGaugeVec(o GaugeOpts, labels ...string) *GaugeVec {
-	m := prometheus.NewGaugeVec(o, labels)
-	switch err := prometheus.Register(m); err.(type) {
-	case nil, prometheus.AlreadyRegisteredError:
-	default:
-		panic(err)
+func (m *Metrics) batchInsertInfluxDB(results *os.File) error {
+	sink := m.writeToInfluxDBSink("results")
+
+	for dec := json.NewDecoder(results); dec.More(); {
+		var me Metric
+		if err := dec.Decode(&me); err != nil {
+			m.re.RecordMessage("failed to decode Metric from results.out: %s", err)
+			continue
+		}
+
+		if err := sink(&me); err != nil {
+			m.re.RecordMessage("failed to process Metric from results.out: %s", err)
+		}
 	}
-	return m
+	return nil
 }
 
-func (*Metrics) NewHistogramVec(o HistogramOpts, labels ...string) *HistogramVec {
-	m := prometheus.NewHistogramVec(o, labels)
-	switch err := prometheus.Register(m); err.(type) {
-	case nil, prometheus.AlreadyRegisteredError:
-	default:
+func (m *Metrics) logSinkJSON(filename string) MetricSinkFn {
+	f, err := m.re.CreateRawAsset(filename)
+	if err != nil {
 		panic(err)
 	}
-	return m
+
+	enc := json.NewEncoder(f)
+	return func(m *Metric) error {
+		return enc.Encode(m)
+	}
 }
 
-func (*Metrics) NewSummaryVec(o SummaryOpts, labels ...string) *SummaryVec {
-	m := prometheus.NewSummaryVec(o, labels)
-	switch err := prometheus.Register(m); err.(type) {
-	case nil, prometheus.AlreadyRegisteredError:
-	default:
-		panic(err)
+func (m *Metrics) writeToInfluxDBSink(measurement string) MetricSinkFn {
```

Comment:
@raulk @coryschwartz I think we have to rethink this method. I don't think we want to send our whole metrics registry into two `measurements`, but rather create a `measurement` for every type of `thing` we want to measure.

Each `metric.Measures` is generally one `measurement` IMO, and since there are 5-6 types (counter, gauge, timer, histogram, etc.) - they have different `fields`.